packages/astro/src/@types/astro.ts

```diff
@@ -1660,12 +1660,47 @@ export type PaginateFunction = (data: any[], args?: PaginateOptions) => GetStati
 
 export type Params = Record<string, string | undefined>;
 
+export type SupportsKind = 'Unsupported' | 'Stable' | 'Experimental' | 'Deprecated';
```

Comment:
Is there a reason we're not hiding `build.split`, `build.excludeMiddleware`, etc, but introducing a guard like this instead?

I was hoping we can support `vercel({ edgeMiddleware: true })` for example, which internally toggles some feature flag, e.g. `supportedFeatures.edgeMiddleware: true` in place of `build.excludeMiddleware: true`. This also makes the feature supported by adapters explicit in the types.

Reply:
> Is there a reason we're not hiding build.split, build.excludeMiddleware, etc, but introducing a guard like this instead?

Do you want to remove them? If so, it would be a breaking change.

> I was hoping we can support `vercel({ edgeMiddleware: true })` for example, which internally toggles some feature flag, e.g. `supportedFeatures.edgeMiddleware: true` in place of `build.excludeMiddleware: true`. This also makes the feature supported by adapters explicit in the types.

The main idea is to tell Astro what an adapter is capable of, and prevent the user from using some adapter/configuration when using x, y or z adapter.

The idea that you suggest can still be implemented. What I want to make clear is that there's a distinction:
- the support map is to tell Astro, "Hey, I - Vercel Adapter - am able to handle the `excludeMiddleware` option.
- your option is more like "Hey, I - user that is using Vercel Adapter - want **to use** this feature from the map.

This means that **wanting to use a feature** doesn't necessarily mean that the adapter is able to support it.

Reply:
Yeah I'm thinking more of no2 that makes more sense to me 🤔 What this becomes is not an Astro feature, but an adapter feature, and for features like "edge middleware" it sounds more of a responsibility/feature of an adapter, rather than Astro.

The feature set I'm thinking off, are a way to tweak Astro to output in a certain way. In most cases, adapters should trying tweaking the astro/vite config before resorting to that.

> Do you want to remove them? If so, it would be a breaking change.

Yeah I think we should remove them. Like above, I don't think they're Astro features, but adapter features. The options are glue code to tweak the Astro output in specific ways. It would be breaking, but worth the newer API design? Just my 2c 😬 

---

I think what this PR has now it's a bit of bouncing responsibilities, and makes it hard to tell what features the adapter supports until you run the adapter (where the error appears)

Reply:
How would we hide `build.split` and `build.excludeMiddleware`? I think we originally did it this way to avoid having another flag to support. But now it makes more sense; since users can set these things anyways don't we want a good way to know if they are going to work? Unless there's a real means to make these "private" to integrations.

Reply:
I still want to make sure that @bluwy 's proposal is unrelated to what I am proposing in this PR.

In this PR, I want to build a common foundation where adapters can tell Astro they support arbitrary features. It's **not** a proposal of how config/options should be passed to Astro.

So far, we have yet to have a standard way to validate certain things. Some example:
- https://github.com/withastro/astro/blob/main/packages/integrations/cloudflare/src/index.ts#L66-L71
- https://github.com/withastro/astro/blob/e5e5cc8e70bd7e842d8cdeb8614552e621e3aea0/packages/astro/src/assets/vite-plugin-assets.ts#L39-L61

That's what this PR wants to address.

Reply:
> How would we hide `build.split` and `build.excludeMiddleware`?

We can have an internal config like `settings._adapterFeatures.split` or `settings.config._adapterFeatures.split`.

> But now it makes more sense; since users can set these things anyways don't we want a good way to know if they are going to work?

I think I'm proposing that users should set these config from the adapter options instead. If it's available in the config types, it would definitely work for the adapter. Perhaps you're talking about the `staticOutput` feature stuff, which in response to Ema:

> In this PR, I want to build a common foundation where adapters can tell Astro they support arbitrary features. It's **not** a proposal of how config/options should be passed to Astro.

Putting it this way, yeah I think that make sense. For features like `staticOutput`, `hybridOuput`, `assets`, I think this PR creates a good way to log that since they are core Astro features.

I think I'm mostly prefer removing the `edgeMiddleware` and `functionsPerPage` features in this PR, since they're not core Astro features, but more of glue logic to make adapters output differently?

Reply:
> I think I'm mostly prefer removing the edgeMiddleware and functionsPerPage features in this PR, since they're not core Astro features, but more of glue logic to make adapters output differently?

You are right, and I agree with you. Although, technically, we still need it. Few examples:

- `edgeMiddleware`: the Vercel adapter can support this feature in "serverless" but doesn't for the "edge". So we need to tell Astro (and the user);
- `functionsPerPage`: we recently implemented this feature - `build.split` - for the Netlify adapter. Although, we implemented it only for the "functions". A few days ago, a user filed an issue because the "edge-functions" didn't support this feature. With this feature map, we could print an error "for free". And eventually, enable this feature once we implement the code to support it.